### DISCUSSION TITLE
Welcome to drafter Discussions!

### DISCUSSION THREAD

**Discussion by acbart** [drafter-edu/drafter#32] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by VictorNorman** [comment]
Hello!  I am Prof. Victor Norman at Calvin University (vtn2@calvin.edu).

I've been looking at Drafter since reviewing the paper for SIGCSE.

I'm playing with it now, trying to implement an assignment we have our students do with guizero: a calculator where you see buttons 1, 2, 3, 4, 5, 6, 7, 8, 9, 0 and  +, -, and = buttons.  When the user clicks a number, the display should update with the number, and keep adding to that number until an operator button is clicked. 

I have some code to show the 10 number buttons, and I've bound them all to a single function. I'd like the function to take in a 2nd argument -- the number of the button that was clicked. SO, I've written code like this:

```
  Span(
                Button("7", numberClicked, [Argument("num", 7)]),
                Button("8", numberClicked, [Argument("num", 8)]),
                Button("9", numberClicked, [Argument("num", 9)]),
                "<p>",
            ),
```

and:

```
@route
def numberClicked(state: State2, num: int):
    state.num1 = num
    print("numberClicked: ", num)
```
but this is really not working. Can you give me an example of how to do what I'm trying to do?  Should I use the Argument list or is there a better way?

Thanks.

**Reply by acbart** [reply]
Sure, I'm happy to help. Here's an example of such a calculator, if I understood what you were looking for. Initially it displays nothing. On every click, it concatenates the given digit to the displayed `num1`.

```python
from drafter import *


@dataclass
class State:
    num1: str


@route
def index(state: State) -> Page:
    return Page(state, [
        Row("Current Value:", state.num1),
        Row(
            Button("7", number_clicked, [Argument("digit", "7")]),
            Button("8", number_clicked, [Argument("digit", "8")]),
            Button("9", number_clicked, [Argument("digit", "9")]),
        )
    ])

@route
def number_clicked(state: State, digit: str) -> Page:
    state.num1 += digit
    return index(state)

start_server(State(""))
```

I suspect the unintuitive part is that the `number_clicked` route needs to return the new `Page` content. You could explicitly replicate the function logic in `index`, but in these cases it's often easiest to just completely reuse the `index` function directly (since Drafter route functions really are just the function they claim to be - the decorator just registers it in a tracking dictionary, it doesn't really make any important modifications). Hence, you can see in my example that the `number_clicked` just calls `index` itself, passing in the modified `state`. It's the kind of thing that feels like overkill (returning an entirely new page for a small update!), but actually works surprisingly okay in practice. 

Also, just as unrelated note: I've been working on version 2.0.0 of Drafter for the past few months. We're actually getting close to launching sometime this Spring. The student-facing API will not change (*should* be fully backwards compatible), but the internals will be vastly improved (both in terms of stability and performance), and there'll be a lot of really interesting new features that I'm will expand what Drafter can do (e.g., working with sensors, supporting a wider set of cool libraries). One relevant feature: you can do partial page updates in Drafter v2 (by returning `Fragment` objects instead of `Page` objects).

**Reply by acbart** [reply]
Oh, I should also mention - I assume you've gone through the [official docs](https://drafter-edu.github.io/drafter/workbook/index.html), and possibly the [examples in this repo](https://github.com/drafter-edu/drafter/tree/main/examples), but I also want to point out the four longer examples (starting with [Ghost Hunter](https://blockpy.cis.udel.edu/assignments/textbook/bakery_textbook?page=bakery_project2_example1_read)) in our Bakery textbook.

**Reply by VictorNorman** [reply]
Thanks!  I've looked through the examples, but didn't see any example that used the Argument list.  Did I miss it somewhere?

**Reply by VictorNorman** [reply]
Also, is `Row` something new? I don't see it the documentation.

**Reply by acbart** [reply]
From the textbook: Ghost Hunt, Protein Lookup, Commenter, and Todo List all use the `arguments` parameter. 

From the `examples/` folder, there's [button_arguments.py](https://github.com/drafter-edu/drafter/blob/main/examples/button_arguments.py) and [shop.py](https://github.com/drafter-edu/drafter/blob/main/examples/shop.py) that immediately jump to mind.

I should move some of these into the official docs website, thanks for pointing out that none of them seem to have any good examples of this. I need to curate the examples on there anyway!

**Reply by acbart** [reply]
> Also, is `Row` something new? I don't see it the documentation.

Yes relatively new, it's something we added just last year, I believe. It's documented [here](https://drafter-edu.github.io/drafter/students/docs.html#Row)
